src/app/clusters/scenes/SceneTableImpl.cpp

```diff
@@ -0,0 +1,680 @@
+                    prev.next = this->next;
+                    ReturnErrorOnFailure(prev.Save(storage));
+                }
+                VerifyOrReturnError(fabric_list.entry_count > 0, CHIP_ERROR_INTERNAL);
+                fabric_list.entry_count--;
+                return fabric_list.Save(storage);
+            }
+            prev                = fabric;
+            fabric.fabric_index = fabric.next;
+        }
+        // Fabric not in the list
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    /// @brief Finds the index where to insert current scene by going through the whole table and looking if the scene is already in
+    /// there. If the target is not in the table, sets idx to the first empty space
+    /// @param target_scene Storage Id of scene to store
+    /// @param idx Index where target or space is found
+    /// @return CHIP_NO_ERROR if managed to find the target scene, CHIP_ERROR_NOT_FOUND if not found and space left
+    ///         CHIP_ERROR_NO_MEMORY if target was not found and table is full
+    CHIP_ERROR Find(SceneStorageId target_scene, SceneIndex & idx)
+    {
+        SceneIndex firstFreeIdx = kUndefinedSceneIndex; // storage index if scene not found
+        uint8_t index           = 0;
+
+        while (index < kMaxScenePerFabric)
+        {
+            if (scene_map[index] == target_scene)
+            {
+                idx = index;
+                return CHIP_NO_ERROR; // return scene at current index if scene found
+            }
+            if (scene_map[index].mEndpointId == kInvalidEndpointId && firstFreeIdx == kUndefinedSceneIndex)
+            {
+                firstFreeIdx = index;
+            }
+            index++;
+        }
+
+        if (firstFreeIdx < kMaxScenePerFabric)
+        {
+            idx = firstFreeIdx;
+            return CHIP_ERROR_NOT_FOUND;
+        }
+
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    CHIP_ERROR Save(PersistentStorageDelegate * storage) override
+    {
+        ReturnErrorOnFailure(Register(storage));
+        return PersistentData::Save(storage);
+    }
+
+    CHIP_ERROR Delete(PersistentStorageDelegate * storage) override
+    {
+        ReturnErrorOnFailure(Unregister(storage));
+        return PersistentData::Delete(storage);
+    }
+
+    CHIP_ERROR SaveScene(PersistentStorageDelegate * storage, const SceneTableEntry & entry)
+    {
+        CHIP_ERROR err;
+        SceneTableData scene(fabric_index, entry.mStorageId, entry.mStorageData);
+        // Look for empty storage space
+
+        err = this->Find(entry.mStorageId, scene.index);
+
+        if (CHIP_NO_ERROR == err)
+        {
+
+            return scene.Save(storage);
+        }
+
+        if (CHIP_ERROR_NOT_FOUND == err) // If not found, scene.index should be the first free index
+        {
+            scene_count++;
+            scene_map[scene.index] = scene.mStorageId;
+            ReturnErrorOnFailure(this->Save(storage));
+
+            return scene.Save(storage);
```

Comment:
If this fails, don't we now have a scene_count that overruns the actual storage?  This seems dubious.  What should the error handling here look like so that we don't end up in inconsistent states?